[Hyunseok Jung] iP

petermonky opened the pull request:
# Monke

> "Ooh ooh, ah ah." – [Japanese Macaque](https://en.wikipedia.org/wiki/Japanese_macaque)

Monke is awesome. It's:

- super fast
- easy to interact with
- banan 🍌

Here are just a couple of its features:

- [x] manage tasks
- [x] that's about it

Let's count to three together:

1. **one**
2. _two_
3. ~~three~~

---

By the way, look at this cool `toString` method I just found 😮:

```java
@Override
public String toString() {
    return "banan";
}
```

Commit:
Add Gradle support

petermonky committed:
Implemented Level-1

petermonky committed:
Implemented Level-2

petermonky committed:
Implemented Level-3

petermonky committed:
Implemented Level-4

petermonky committed:
Implemented testing using I/O redirection

petermonky committed:
Implemented Level-5

petermonky committed:
Minor spacing fix

petermonky committed:
According spacing fix for testing file

petermonky committed:
Implemented Level-6

petermonky committed:
Implemented enums for clearer logic

petermonky committed:
Personality core attached

petermonky committed:
Implement Level-7

Duke writes and saves tasks to an external file.

Data is stored locally on the user's device.

petermonky committed:
Implement Level-8

Duke understands LocalDateTime.

Time of deadline/event is printed in a different format.

petermonky committed:
Merge branch 'branch-Level-7'

petermonky committed:
Merge branch 'branch-Level-8'

petermonky committed:
Duke: improve OOP integration

More compartmentalisation of components.

petermonky committed:
Duke: divide into packages

Package classes categorically.

petermonky committed:
Parser: change access modifier

petermonky committed:
Test: add tests for tasks

Implement JUnit tests for Task subclasses.

petermonky committed:
Duke: add JavaDoc for classes

Class and method JavaDoc headers.

petermonky committed:
Duke: follow coding standard

Improve readability with line breaks.

petermonky committed:
Find Command: implement find feature

Search for task using keyword.

petermonky committed:
Duke: resolve merge conflict
Merge branch-A-CodingStandard

petermonky committed:
Duke: resolve merge conflict

Merge branch-Level-9

tanyutao544 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,46 @@
+import duke.logic.Storage;
+import duke.logic.TaskList;
+import duke.logic.Ui;
+
```
wouldn't adding javadoc in this class help in understanding what the program is doing? 🤔

tanyutao544 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public void run() {
+        this.ui.showIntro();
```
is "this." here needed?

tanyutao544 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,46 @@
+            try {
+                String fullCommand = ui.readCommand();
+                Command command = Parser.parse(fullCommand);
+                hasExited = !command.execute(this.taskList, this.ui, this.storage);
```
is "this." here needed as well?

tanyutao544 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,24 @@
+     * @return boolean value that determines process termination.
+     * @throws DukeException If operation is unsuccessful.
+     */
+    public abstract boolean execute(TaskList taskList, Ui ui, Storage storage) throws DukeException;
```
I like how you compacted execute to return the program termination state, instead of implementing a separate getter function 😎

tanyutao544 commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,59 @@
+            task.markAsUndone();
+            ui.showMessage("TASK UNDONE:\n" + task);
+        }
+        storage.writeToFile(taskList);
```
This is nice, incase the program exits unexpectedly, your previous commands still save. Good thinking.

tanyutao544 commented on src/main/java/duke/logic/Parser.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    /**
+     * Parses the time portion of a command.
```
Is the method description here correct?

tanyutao544 commented on src/main/java/duke/logic/Parser.java:
```diff
@@ -0,0 +1,137 @@
+            case MARK:
+                command = new MarkCommand(parseIndex(splitInput[1]), true);
+                break;
+            case UNMARK:
```
I think it is possible to compact this "unmark" case with "mark" case
E.G. 
case MARK:
case UNMARK:
    command = new ......
    break;
Could be something to consider

tanyutao544 commented on src/main/java/duke/logic/Parser.java:
```diff
@@ -0,0 +1,137 @@
+ */
+public class Parser {
+    private enum CommandType {
+        BYE, LIST, MARK, UNMARK, DELETE, TODO, DEADLINE, EVENT, FIND, CLEAR
```
Smart use of enums here

tanyutao544 commented on src/main/java/duke/logic/Storage.java:
```diff
@@ -0,0 +1,120 @@
+     * @throws DukeException If read from local file is unsuccessful.
+     */
+    public ArrayList<Task> readFromFile() throws DukeException {
+        ArrayList<Task> tasks = new ArrayList<>();
```
Couldn't this be replaced with your custom tasklist class?

tanyutao544 reviewed (commented):
Nicely structured project, almost zero things to fix

xSaints19x commented on src/main/java/duke/logic/Storage.java:
```diff
@@ -0,0 +1,120 @@
+    public ArrayList<Task> readFromFile() throws DukeException {
+        ArrayList<Task> tasks = new ArrayList<>();
+
+        try {
+            Scanner sc = new Scanner(this.file);
+
+            while (sc.hasNextLine()) {
+                String line = sc.nextLine();
+                String[] breakdown = line.split(" \\| ");
+                String command = breakdown[0];
+
+                try {
+                    TaskType taskType = TaskType.valueOf(command);
+
+                    switch (taskType) {
+                    case T:
+                        tasks.add(new Todo(breakdown[2],
+                                Boolean.parseBoolean(breakdown[1])));
+                        break;
+                    case E:
+                        tasks.add(new Event(breakdown[2],
+                                Boolean.parseBoolean(breakdown[1]),
+                                LocalDateTime.parse(breakdown[3])));
+                        break;
+                    case D:
+                        tasks.add(new Deadline(breakdown[2],
+                                Boolean.parseBoolean(breakdown[1]),
+                                LocalDateTime.parse(breakdown[3])));
+                        break;
+                    }
+                } catch (IllegalArgumentException e) {
+                    throw new DukeException("INVALID TASK TYPE");
+                }
+            }
+
+            return tasks;
+        } catch (IOException e) {
+            throw new DukeException("INVALID FILE PATH");
+        }
```
```suggestion
        try {
            Scanner sc = new Scanner(this.file);
        } catch (IOException e) {
            throw new DukeException("INVALID FILE PATH");
        }

        while (sc.hasNextLine()) {
            String line = sc.nextLine();
            String[] breakdown = line.split(" \\| ");
            String command = breakdown[0];

            try {
                TaskType taskType = TaskType.valueOf(command);

                switch (taskType) {
                case T:
                    tasks.add(new Todo(breakdown[2],
                            Boolean.parseBoolean(breakdown[1])));
                    break;
                case E:
                    tasks.add(new Event(breakdown[2],
                            Boolean.parseBoolean(breakdown[1]),
                            LocalDateTime.parse(breakdown[3])));
                    break;
                case D:
                    tasks.add(new Deadline(breakdown[2],
                            Boolean.parseBoolean(breakdown[1]),
                            LocalDateTime.parse(breakdown[3])));
                    break;
                }
            } catch (IllegalArgumentException e) {
                  throw new DukeException("INVALID TASK TYPE");
            }
        }
        return tasks;
         
```
Perhaps placing the try-catch block above the while-loop will be better to avoid _arrowhead_ style code?

xSaints19x commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Boolean flag that determines whether to mark or unmark task.
+     */
+    private final boolean isMark;
```
```suggestion
    private final boolean isMarked;
```
Perhaps the above suggestion would be good here?

xSaints19x reviewed (commented):
Overall great job well done. 👍
Just some minor cosmetic changes requested.

petermonky committed:
Merge branch 'origin/add-gradle-support'

petermonky committed:
Gradle: enable checkstyle

petermonky committed:
Duke: improve coding style

petermonky committed:
Duke: implement GUI

Duke runs in GUI mode.

Makes use of JavaFX.

petermonky committed:
Duke: remove unused method

petermonky committed:
Merge branch 'branch-Level-10'

petermonky committed:
Duke: improve find command

Find command accepts varargs.

Look for multiple keywords.

petermonky committed:
Duke: change GUI appearance

petermonky committed:
Duke: handle invalid index

Oddcorner commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    @Override
+    public void start(Stage stage) {
```
Could this method be shortened by abstracting the initialisation of different items into methods like `setStage`, `setScrollPane`, `setAnchorPane` and calling them here instead?

Oddcorner commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,55 @@
+    public String execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        taskList.add(this.task);
+        storage.writeToFile(taskList);
+        String output = "TASK ADDED:\n"
+                + task + "\n"
+                + taskList.size() + " TASK(S) NOW.";
```
`output` could be defined as a private field of the AddCommnad class instead of being defined here. This will allow for a Single Layer of Abstraction (SLAP) in this method and makes it easier to find where to change the output message if needed in the future. What do you think?

Oddcorner commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,200 @@
+            Timer timer = new Timer();
+            Command command = Parser.parse(fullCommand);
+            if (command.isExitCommand()) {
+                timer.schedule(new TimerTask() {
+                    @Override
+                    public void run() {
+                        Platform.exit();
+                        System.exit(0);
+                    }
+                }, 1000);
```
Could this instantiation be done outside to avoid deep nesting?

Oddcorner commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,48 @@
+    public String execute(TaskList taskList, Ui ui, Storage storage) {
+        TaskList newTaskList = taskList.filter(this.keywords);
+        String output = "MATCHING TASKS HERE:" + newTaskList;
+        ui.showMessage("MATCHING TASKS HERE:" + newTaskList);
```
Would doing this be more appropriate as you would need to remember to edit the message here as well if there are changes to the message otherwise?
```suggestion
        ui.showMessage(output);
```

Oddcorner commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,72 @@
+            throw new DukeException("INVALID INDEX");
+        }
+
+        if (isMarkCommand) {
+            task.markAsDone();
+            output = "TASK DONE:\n" + task;
+        } else {
+            task.markAsUndone();
+            output = "TASK UNDONE:\n" + task;
+        }
```
Nice use of a final boolean to make the Mark and Unmark command into a single class. 👍

Oddcorner commented on src/main/java/duke/gui/DialogueBox.java:
```diff
@@ -0,0 +1,35 @@
+     * @param l  Text to be displayed.
+     * @param iv Window to contain text.
+     */
+    public DialogueBox(Label l, ImageView iv, boolean isUser) {
```
Would using more descriptive names for `Label` and `ImageView` be preferred? 
```suggestion
    public DialogueBox(Label message, ImageView picture, boolean isUser) {
```

Oddcorner reviewed (commented):
Overall nice code quality, good use of enums and nice OOP principles applied. 👍 

Just take note of repeated code, and consider splitting longer code blocks into shorter simpler methods to improve readability!

petermonky committed:
Duke: improve GUI

Colour-code messages

Add circular frame to images

petermonky committed:
build.gradle: change class name

Correctly compile main Duke programme.

petermonky committed:
Duke: add assert statement

petermonky committed:
Duke: improve code quality

Refer to PR comments from peers.

petermonky committed:
Merge pull request #1 from petermonky/branch-A-Assertions

Duke: add assert statement

petermonky committed:
Merge branch 'master' into branch-A-CodeQuality

petermonky committed:
Merge pull request #2 from petermonky/branch-A-CodeQuality

Duke: improve code quality

petermonky committed:
Duke: add undo command

Rudimentary implementation of undo.

Does not work with un/mark commands.

petermonky committed:
DeleteCommand: fix undo handling

petermonky committed:
UndoCommand: fix formatting

petermonky committed:
Duke: refactor GUI components

Compartmentalise initialisation and setup of Duke GUI components.

Implement ExitDukeTimerTask to specifically handle programme termination
 of Duke GUI.

petermonky committed:
Duke: add more descriptive comments

petermonky committed:
TaskList: remove unnecessary assert statement

petermonky committed:
Duke: add Ui.png

petermonky committed:
DialogueBox: adjust font and image size

petermonky committed:
Set theme jekyll-theme-minimal

petermonky committed:
Create CNAME

petermonky committed:
Delete CNAME

petermonky committed:
Set theme jekyll-theme-hacker

petermonky committed:
Set theme jekyll-theme-tactile

petermonky committed:
Set theme jekyll-theme-cayman

petermonky committed:
README: update user guide

petermonky committed:
Merge remote-tracking branch 'origin/master'

petermonky committed:
Duke: debug undo command

Correctly perform undo with tasks.

petermonky committed:
Duke: refactor for checkstyle
